### DISCUSSION TITLE
i18n primitive type overhaul

### DISCUSSION THREAD

**Discussion by milomg** [solidjs-community/solid-primitives#160] [Ideas]
I've been playing around with typescript-eslint in solid-site and it complains that template solidI18n returns any instead of a string (which makes all operations with it untyped). 

I think we should expect that the translation is just a deeply nested object with strings as leaves, and change the translate function to only return a string

The only downside I can see to this is not allowing easy dynamic arrays. For example, in solid-site we currently have an array for each of the nav items. However, I believe this shouldn't be done with the i18n provider, and instead done with some other service if necessary.

**Comment by milomg** [comment]
```ts
type Nested = Record<string, Nested | string>;

[
  template: (key: string, params?: Record<string, string>, defaultValue?: string) => string,
  actions: {
    add(lang: string, table: Nested): void;
    locale: (lang?: string) => string;
    dict: (lang: string) => Record<string, Nested>;
  }
]

**Reply by davedbase** [reply]
I don't see an issue with the nesting fix. Unless I'm missing something I don't think this changes the other aspects of the package too dramatically.

**Comment by mb21** [comment]
https://primitives.solidjs.community/package/i18n is currently on stage 3... should this status be reconsidered? 🤔

**Comment by thetarnav** [comment]
An overhaul: https://github.com/solidjs-community/solid-primitives/pull/510
